File: ISP/ISP.java
package SOLID;
interface RemoteControl {
    void turnOn();
    void turnOff();
    void record(); // not needed by all
}
class TV implements RemoteControl {
    public void turnOn() {
    //logic
    }
    public void turnOff() {
    //logic
    }

    public void record() {
        throw new UnsupportedOperationException();
    }
}
class Camera implements RemoteControl {
    public void turnOn() {
        //logic
    }
    public void turnOff() {
        //logic
    }
    public void record() {
        //logic
    }
}

interface Switchable {
    void turnOn();
    void turnOff();
}
interface Recordable {
    void record();
}
class TVGood implements Switchable {
    public void turnOn() {
        //logic
    }
    public void turnOff() {
        //logic
    }
}
class CameraFixed implements Switchable, Recordable {
    public void turnOn() {
        //logic
    }
    public void turnOff() {
        //logic
    }

    public void record() {
        //logic
    }
}

public class ISP {
}

File: ISP/ISP.ts
interface RemoteControl {
    turnOn(): void;
    turnOff(): void;
    record(): void;
}
class TV implements RemoteControl {
    turnOn() {
//LOGIC
}
    turnOff() {
//logic
    }
    record() {
        throw new Error("TV cannot record");
    }
}

class Camera implements RemoteControl {
    turnOn() {
        //logic
    }

    turnOff() {
        //logic
    }

    record() {
        //logic
    }
}

//good code
interface Switchable {
    turnOn(): void;
    turnOff(): void;
}
interface Recordable {
    record(): void;
}
class TVGood implements Switchable {
    turnOn() {
        //logic
    }
    turnOff() {
        //logic
    }
}
class CameraGood implements Switchable, Recordable {
    turnOn() {
    //logic
    }
    turnOff() {
        //logic
    }
    record() {
        //logic
    }
}
